### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by samuelkarp**
@aaithal @richardpen @nmeyerhans @kiranmeduri 

### Summary
[Build status](https://ci.appveyor.com/project/samuelkarp/amazon-ecs-agent/branch/windows)

This pull request gets the ECS agent working on Windows.  This is a work in progress and there are a number of things that don't work or don't work very well.  I'll be continuing to work on this until it's ready to merge.

### Implementation details
Much of the code works on Windows without issue.  However, a few things did need to change:
* I needed to add support for Windows named pipes to [go-dockerclient](https://github.com/fsouza/go-dockerclient/pull/579) in order to communicate with the Docker daemon
* I've migrated our state-saving mechanism that uses POSIX filesystem semantics for atomicity on Linux to use the Windows registry for atomicity
* I've had to disable our debug handler that uses `SIGUSR1` on Linux; we can migrate to using a Win32 event [like Docker does](https://github.com/docker/docker/blob/v1.13.0-rc1/daemon/debugtrap_windows.go#L15), but I haven't done this yet
* I've accounted for case-insensitivity for file paths on Windows (though there may still be bugs lurking here)
* I've worked around a [bug in Docker](https://github.com/docker/docker/issues/27601) involving the output of `docker inspect`
* Additional setup is necessary on the host in order to use IAM roles for tasks; these are run as part of `misc/windows-deploy/agent.ps1`
* I've added some scripts to run integ and functional test with the setup that's required in the absence of `make` for Windows
* I've added Windows-specific default configuration

### Testing
I've run a few simple task definitions for Windows involving multiple containers and volume mounts.  Unit tests are passing on Windows: [Build status](https://ci.appveyor.com/project/samuelkarp/amazon-ecs-agent/branch/windows).  Many integ tests and functional tests have also been ported to Windows and pass.

- [x] Builds (`make release` on Linux and `go build .\agent` on Windows)
- [x] Unit tests (`make test` on Linux and `go test -timeout=25s .\agent\...` on Windows) pass
- [x] Integration tests (`make run-integ-tests` on Linux and `.\scripts\run-integ-tests.ps1` on Windows) pass
- [x] Functional tests (`make run-functional-tests` on Linux and `.\scripts\run-functional-tests.ps1` on Windows) pass

New tests cover the changes: yes

### Description for the changelog
Feature - Support for Docker on Windows Server 2016

### Licensing
This contribution is under the terms of the Apache 2.0 License: yes (Amazon employee)

### CODE REVIEW

File: agent/statemanager/state_manager_windows.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		seelog.Errorf("Error saving state; could not create file to save state: %v", err)
+		return err
+	}
+	defer dataFile.Close()
+	_, err = dataFile.Write(data)
+	if err != nil {
+		seelog.Errorf("Error saving state; could not write to file to save state: %s %v ", dataFile.Name(), err)
+		return err
+	}
+	err = dataFile.Sync() // this calls FlushFileBuffers, see https://golang.org/src/syscall/syscall_windows.go#L523
```
Comment (khannag) [inline agent/statemanager/state_manager_windows.go]: FlushFileBuffers has performance implications per msdn. It might be useful to characterize the overhead (time taken to sync) and see if it entails switching to other options like unbuffered IO. This is good for now but likely needs revisiting in future.

Comment (samuelkarp) [inline agent/statemanager/state_manager_windows.go]: Sure.  I can add a TODO here if you'd like.

File: agent/api/const_windows.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package api
+
+const OSType = "windows"
```
Comment (samuelkarp) [inline agent/api/const_windows.go]: This registers an attribute called `ecs.os-type` that will show up when you call `DescribeContainerInstances`.

File: agent/config/config_windows.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		ReservedPortsUDP:            []uint16{},
+		DataDir:                     filepath.Join(ecsRoot, "data"),
+		DisableMetrics:              false,
+		ReservedMemory:              0,
+		AvailableLoggingDrivers:     []dockerclient.LoggingDriver{dockerclient.JsonFileDriver},
+		TaskCleanupWaitDuration:     DefaultTaskCleanupWaitDuration,
+		DockerStopTimeout:           DefaultDockerStopTimeout,
+		CredentialsAuditLogFile:     filepath.Join(ecsRoot, defaultCredentialsAuditLogFile),
+		CredentialsAuditLogDisabled: false,
+		ImageCleanupDisabled:        false,
```
Comment (khannag) [inline agent/config/config_windows.go]: Not sure if I understand this fully, but does this mean if a customer never changes this setting, that the container images will never be cleaned up ? The local disk may eventually run out storage space since these images (especially on Windows) have a sizable disk footprint.

Comment (samuelkarp) [inline agent/config/config_windows.go]: disabled = false means that it is enabled by default.  There's a bit of technical debt in our config package that means our defaults must be false in order to be overridden, so anything that is enabled by default has a config field for "disabled" instead of enabled.

tl;dr: Image cleanup is enabled by default on both Windows and Linux.

File: agent/functional_tests/tests/functionaltests_windows_test.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+const (
+	waitTaskStateChangeDuration     = 2 * time.Minute
+	waitMetricsInCloudwatchDuration = 4 * time.Minute
+	awslogsLogGroupName             = "ecs-functional-tests"
+)
+
+// TestRunManyTasks runs several tasks in short succession and expects them to
+// all run.
+func TestRunManyTasks(t *testing.T) {
+fmt.Println("starting")
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_windows_test.go]: there are a number of these in this file. You might wanto to get rid of this. Or, use log instead

File: agent/functional_tests/tests/functionaltests_windows_test.go [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+	numToRun := 1
+	tasks := []*TestTask{}
+	attemptsTaken := 0
+  fmt.Println("get td")
+	td, err := GetTaskDefinition("simple-exit-windows")
+	if err != nil {
+		t.Fatalf("Get task definition error: %v", err)
+	}
+	fmt.Println("Starting Tasks")
+  t.Logf("Starting Tasks")
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_windows_test.go]: Weird tabbing?

File: agent/functional_tests/tests/functionaltests_windows_test.go [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package functional_tests
+
+import (
+	//"bytes"
+	"fmt"
+	//"io/ioutil"
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_windows_test.go]: delete unused imports?

File: agent/functional_tests/util/utils_windows.go [hunk 7]
```diff
@@ -0,0 +1,778 @@
...
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+	"os/exec"
+  "golang.org/x/sys/windows/registry"
```
Comment (aaithal) [inline agent/functional_tests/util/utils_windows.go]: weird tabbing..

File: agent/functional_tests/util/utils_windows.go [hunk 8]
```diff
@@ -0,0 +1,778 @@
...
+	if err != nil {
+		t.Fatal("Could not create temp dir for test")
+	}
+	logdir := filepath.Join(agentTempdir, "log")
+	datadir := filepath.Join(agentTempdir, "data")
+	os.Setenv("ECS_LOGFILE", logdir + "/log.log")
+	os.Setenv("ECS_DATADIR", datadir)
+	valueName := fmt.Sprintf("test_path_%d", time.Now().UnixNano())
+	fmt.Println("Registry location",valueName)
+  os.Setenv("ZZZ_I_KNOW_SETTING_TEST_VALUES_IN_PRODUCTION_IS_NOT_SUPPORTED", valueName)
```
Comment (aaithal) [inline agent/functional_tests/util/utils_windows.go]: weird tabbing (here and at other lines in this file).. also `fmt.Println()`s

File: agent/functional_tests/util/utils_windows.go [hunk 9]
```diff
@@ -0,0 +1,778 @@
...
+	resp, err := ECS.StartTask(&ecs.StartTaskInput{
+		Cluster:            &agent.Cluster,
+		ContainerInstances: cis,
+		TaskDefinition:     &taskDefinition,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("here")
```
Comment (aaithal) [inline agent/functional_tests/util/utils_windows.go]: remove this other `fmt.Println()`s?

File: agent/functional_tests/tests/functionaltests_windows_test.go (hunk 6 above)
Comment (samuelkarp) [inline agent/functional_tests/tests/functionaltests_windows_test.go]: This is fixed in a later commit.

File: agent/functional_tests/tests/functionaltests_windows_test.go (hunk 4 above)
Comment (samuelkarp) [inline agent/functional_tests/tests/functionaltests_windows_test.go]: This is fixed in a later commit.

File: agent/functional_tests/tests/functionaltests_windows_test.go (hunk 5 above)
Comment (samuelkarp) [inline agent/functional_tests/tests/functionaltests_windows_test.go]: This is fixed in a later commit.

File: agent/functional_tests/util/utils_windows.go (hunk 7 above)
Comment (samuelkarp) [inline agent/functional_tests/util/utils_windows.go]: This is fixed in a later commit.

File: agent/functional_tests/util/utils_windows.go (hunk 8 above)
Comment (samuelkarp) [inline agent/functional_tests/util/utils_windows.go]: I'll change this `fmt.Println` to `t.Log`.  The tabbing is fixed in a later commit.

File: agent/functional_tests/util/utils_windows.go (hunk 9 above)
Comment (samuelkarp) [inline agent/functional_tests/util/utils_windows.go]: This is fixed in a later commit.

File: agent/functional_tests/util/utils_windows.go [hunk 10]
```diff
@@ -215,564 +115,24 @@ func (agent *TestAgent) StopAgent() error {
...
-	fmt.Println("Container InstanceArn:", agent.ContainerInstanceArn)
-	agent.Cluster = localMetadata.Cluster
-	if localMetadata.Version != "" {
-		versionNumberRegex := regexp.MustCompile(` v(\d+\.\d+\.\d+) `)
-		versionNumberStr := versionNumberRegex.FindStringSubmatch(localMetadata.Version)
-		if len(versionNumberStr) == 2 {
-			agent.Version = string(versionNumberStr[1])
-		}
+func (agent *TestAgent) StartAgent() error {
+	agentInvoke := exec.Command(".\\agent.exe")
```
Comment (aaithal) [inline agent/functional_tests/util/utils_windows.go]: Just a note here that we need a windows readme for functional tests

Comment (samuelkarp) [inline agent/functional_tests/util/utils_windows.go]: [This README](https://github.com/samuelkarp/amazon-ecs-agent/blob/8d78cecbeeebcdb1ab818f89cfe48ae76dbf5164/agent/functional_tests/README.md#windows-setup) covers Windows functional tests.  We do need changes to the general README for Windows though.

File: misc/windows-iam/Setup_Iam.ps1 [hunk 11]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2014-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved
```
Comment (samuelkarp) [inline misc/windows-iam/Setup_Iam.ps1]: The networking changes required for IAM roles to work on Windows are significantly different from that on Linux.  We will need to document how this works in more detail, but may not get to it before merging this.

File: misc/windows-listen80/listen80.go [hunk 12]
```diff
@@ -0,0 +1,29 @@
+// +build !windows
+
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (aaithal) [inline misc/windows-listen80/listen80.go]: nit: %s/2015/2016/

File: agent/sighandlers/debug_handler_windows.go [hunk 13]
```diff
@@ -0,0 +1,19 @@
+// +build windows
+
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (richardpen) [inline agent/sighandlers/debug_handler_windows.go]: should this be 2014-2016

Comment (samuelkarp) [inline agent/sighandlers/debug_handler_windows.go]: Fixed.

File: agent/functional_tests/util/utils_windows.go (hunk 8 above)
Comment (samuelkarp) [inline agent/functional_tests/util/utils_windows.go]: Fixed.

**Comment by aaithal** [comment]
Went through the code once. Will probably go through it again. Overall, this looks really good. No major comments. Will update the PR after going through it again.

File: agent/functional_tests/util/utils_windows.go (hunk 10 above)
Comment (aaithal) [inline agent/functional_tests/util/utils_windows.go]: Awesome, thanks!

File: misc/windows-listen80/listen80.dockerfile [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+#
+#	http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+FROM microsoft/windowsservercore:latest
+ADD listen80.exe listen80.exe
```
Comment (richardpen) [inline misc/windows-listen80/listen80.dockerfile]: I guess you should delete one listen80.exe?

Comment (richardpen) [inline misc/windows-listen80/listen80.dockerfile]: thanks

File: misc/windows-iam/ec2.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (richardpen) [inline misc/windows-iam/ec2.go]: s/2015/2016/

Comment (samuelkarp) [inline misc/windows-iam/ec2.go]: Fixed.

File: agent/api/task_windows.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+	task.downcaseAllVolumePaths()
+}
+
+// downcaseAllVolumePaths forces all volume paths (host path and container path)
+// to be lower-case.  This is to account for Windows case-insensitivity and the
+// case-sensitive string comparison that takes place elsewhere in the code.
+func (task *Task) downcaseAllVolumePaths() {
+	for _, volume := range task.Volumes {
+		if hostVol, ok := volume.Volume.(*FSHostVolume); ok {
+			hostVol.FSSourcePath = strings.ToLower(hostVol.FSSourcePath)
```
Comment (samuelkarp) [inline agent/api/task_windows.go]: Yes, we can add a TODO, but I'd prefer that validation happen when you call `RegisterTaskDefinition` rather than asynchronously when you try to run a task.  I'll note this internally on the list of validations that needs to be added.

File: agent/api/task.go [hunk 17]
```diff
@@ -397,9 +397,9 @@ func (task *Task) dockerPortMap(container *Container) map[docker.Port][]docker.P
 		dockerPort := docker.Port(strconv.Itoa(int(portBinding.ContainerPort)) + "/" + portBinding.Protocol.String())
 		currentMappings, existing := dockerPortMap[dockerPort]
 		if existing {
-			dockerPortMap[dockerPort] = append(currentMappings, docker.PortBinding{HostIP: "0.0.0.0", HostPort: strconv.Itoa(int(portBinding.HostPort))})
+			dockerPortMap[dockerPort] = append(currentMappings, docker.PortBinding{HostIP: portBindingHostIP, HostPort: strconv.Itoa(int(portBinding.HostPort))})
```
Comment (samuelkarp) [inline agent/api/task.go]: Yes, it's a limitation of Windows Net NAT with Docker.

**Comment by samuelkarp** [comment]
@nrdlngr I've updated the README to account for Windows.  Would you mind taking a look?

**Comment by samuelkarp** [comment]
@aaithal @khannag @richardpen @kiranmeduri @nmeyerhans 

I think this is pretty much ready, barring anything new we discover in testing.  Can you take a look?

File: misc/windows-iam/Setup_Iam.ps1 [hunk 18]
```diff
@@ -11,10 +11,40 @@
...
 
-Invoke-Expression "${PSScriptRoot}\loopback.ps1"
-Invoke-Expression "${PSScriptRoot}\hostsetup.ps1"
-Invoke-Expression "docker build -t amazon/amazon-ecs-credential-proxy --file ${PSScriptRoot}\credentialproxy.dockerfile ${PSScriptRoot}"
+$ErrorActionPreference = 'Stop'
 
+# Download devcon binary
+$devconZipUri = "https://s3.amazonaws.com/amazon-ecs-devcon/devcon-3a3f98688359aa1b905147831f910218.zip"
+$devconZipMD5Uri = "$devconZipUri.md5"
+$devconZipFile = "${PSScriptRoot}\devcon.zip"
```
Comment (samuelkarp) [inline misc/windows-iam/Setup_Iam.ps1]: > This suggests that for the purpose of tests, we are copying devcon.exe to the windows-iam folder.

Yes, that's correct.

> is it packaged as part of the ecs agent zip ?

Yes, see [here](https://github.com/samuelkarp/amazon-ecs-agent/blob/dc3924e4835c9e113beadc144c02a1559d7d59db/scripts/stage#L87-L101).

> Other than this, if the below re-organization has been tested, then it looks good to go.

Yep, it's passing for me.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 19]
```diff
@@ -15,18 +15,25 @@ Please provide the following information:
...
-- [ ] Builds (`make release`)
-- [ ] Unit tests (`make test`) pass
-- [ ] Integration tests (`make run-integ-tests`) pass
-- [ ] Functional tests (`make run-functional-tests`) pass
+- [ ] Builds on Linux (`make release`)
+- [ ] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
+- [ ] Unit tests on Linux (`make test`) pass
+- [ ] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
+- [ ] Integration tests on Linux (`make run-integ-tests`) pass
+- [ ] Integratio tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
```
Comment (aaithal) [inline .github/PULL_REQUEST_TEMPLATE.md]: typo: Integration

File: README.md [hunk 20]
```diff
@@ -44,22 +44,49 @@ $ docker run --name ecs-agent \
...
+
+```powershell
+PS C:\> # Set up directories the agent uses
+PS C:\> New-Item -Type directory -Path $ProgramFiles\Amazon\ECS
+PS C:\> New-Item -Type directory -Path $ProgramData\Amazon\ECS
+PS C:\> # Set up configuration
+PS C:\> $ecsExeDir = "$env:ProgramFiles\Amazon\ECS"
+PS C:\> [Environment]::SetEnvironmentVariable("ECS_CLUSTER", "my-windows-cluster", "Machine")
+PS C:\> # Set this environment variable to "true" to enable IAM roles.  Note that enabling IAM roles will make port 80 unavailable for tasks.
+PS C:\> [Environment]::SetEnvironmentVariable("ECS_ENABLE_TASK_IAM_ROLE", "false", "Machine")
```
Comment (aaithal) [inline README.md]: Should we add/set `ECS_LOGFILE` here as well?

Comment (samuelkarp) [inline README.md]: Yes, done

File: scripts/ec2-stage [hunk 21]
```diff
@@ -25,10 +25,19 @@ SOURCE_BUCKET=""
...
-
+SUPPORTED_OSES=("linux" "windows")
+TARGET_OS_LIST=${SUPPORTED_OSES[@]}
 
 source $(dirname "${0}")/publishing-common.sh
 
+supported_os() {
+	local os=$1
+	for entry in ${SUPPORTED_OSES[@]} ; do
+		test "$os" == "$entry" && return $(true)
```
Comment (samuelkarp) [inline scripts/ec2-stage]: I'm not opposed to that but I think @nmeyerhans had an opinion here.

Comment (nmeyerhans) [inline scripts/ec2-stage]: IMO the addition of punctuation where it's not necessary detracts from the readability of the code. The [Google bash styleguide](https://google.github.io/styleguide/shell.xml#Variable_expansion) describes this topic as "too controversial for a mandatory regulation", so I suspect that we'll see different styles based on personal preference for as long as we're writing shell. However, the argument for consistency is valid, so I'm OK with the change.

Comment (richardpen) [inline scripts/ec2-stage]: One more `script/build`, thanks

Comment (samuelkarp) [inline scripts/ec2-stage]: Done

File: agent/statemanager/state_manager_windows.go [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+	"path/filepath"
+
+	"github.com/cihub/seelog"
+
+	"golang.org/x/sys/windows/registry"
+)
+
+const (
+	ecsDataFileRootKey   = registry.LOCAL_MACHINE
+	ecsDataFileKeyPath   = `SOFTWARE\Amazon\ECS Agent\State File`
```
Comment (aaithal) [inline agent/statemanager/state_manager_windows.go]: Can we get rid of the space character here in the [commit](b23f949d5bb0c1345ef7f9bf3e039067c0ade0e2)?

File: agent/statemanager/state_manager_windows.go [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+file to load.
+*/
+
+func (manager *basicStateManager) readFile() ([]byte, error) {
+	manager.savingLock.Lock()
+	defer manager.savingLock.Unlock()
+	path, err := manager.getPath()
+	if err != nil {
+		if err == registry.ErrNotExist {
+			// Happens on the first run; not a real error
```
Comment (aaithal) [inline agent/statemanager/state_manager_windows.go]: Might be useful to log something here ([commit](b23f949d5bb0c1345ef7f9bf3e039067c0ade0e2)).

File: agent/statemanager/state_manager_windows.go [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+			return nil, nil
+		}
+		return nil, err
+	}
+	file, err := os.Open(filepath.Clean(path))
+	if err != nil {
+		if os.IsNotExist(err) {
+			// Happens every first run; not a real error
```
Comment (aaithal) [inline agent/statemanager/state_manager_windows.go]: Might be useful to log something here as well ([commit](b23f949d5bb0c1345ef7f9bf3e039067c0ade0e2)).

File: agent/config/config_windows.go [hunk 25]
```diff
@@ -24,16 +24,40 @@ import (
 
 const (
 	defaultCredentialsAuditLogFile = `log\audit.log`
-	httpPort                       = 80
+	// When using IAM roles for tasks on Windows, the credential proxy consumes port 80
+	httpPort = 80
```
Comment (aaithal) [inline agent/config/config_windows.go]: nit: lint would dictate that this starts with `httpPort specifies the credentials ...`

File: agent/statemanager/state_manager_windows.go (hunk 22 above)
Comment (aaithal) [inline agent/statemanager/state_manager_windows.go]: That's fine. We can leave it as it is.

File: agent/statemanager/state_manager_windows.go (hunk 23 above)
Comment (samuelkarp) [inline agent/statemanager/state_manager_windows.go]: I don't think it's necessary; this is normal and expected.

File: agent/statemanager/state_manager_windows.go (hunk 24 above)
Comment (samuelkarp) [inline agent/statemanager/state_manager_windows.go]: I don't think it's necessary; this is normal and expected.

File: agent/config/config_windows.go (hunk 25 above)
Comment (samuelkarp) [inline agent/config/config_windows.go]: lint only enforces this for public things (consts, variables, structs, struct fields), since there's no godoc on private ones.  I can change this if you feel strongly.
